xdebug_com.c

```diff
@@ -25,165 +25,205 @@
-		if (host_info == 0) {
-			/* Error: unknown host */
-			return -1;
-		}
-		*in = *((struct in_addr *) host_info->h_addr);
-	}
-	return 0;
-}
-/* }}} */
-
 int xdebug_create_socket(const char *hostname, int dport)
 {
-	struct sockaddr    sa;
-	struct sockaddr_in address;
-	int                sockfd;
-	int                status;
-	struct timeval     timeout;
-	int                actually_connected;
-	socklen_t          size = sizeof(sa);
+	struct addrinfo            hints;
+	struct addrinfo            *remote;
+	struct addrinfo            *ptr;
+	int                        status;
+	int                        sockfd;
+	int                        sockerror;
+	char                       sport[10];
+	struct timeval             timeout;
+	int                        actually_connected;
+	struct sockaddr_in6        sa;
+	socklen_t                  size = sizeof(sa);
 #if WIN32|WINNT
-	WORD               wVersionRequested;
-	WSADATA            wsaData;
-	char               optval = 1;
-	const char         yes = 1;
-	u_long             no = 0;
+	WORD                       wVersionRequested;
+	WSADATA                    wsaData;
+	char                       optval = 1;
+	const char                 yes = 1;
+	u_long                     no = 0;
 
 	wVersionRequested = MAKEWORD(2, 2);
 	WSAStartup(wVersionRequested, &wsaData);
 #else
-	long               optval = 1;
+	long            optval = 1;
 #endif
-	memset(&address, 0, sizeof(address));
-	lookup_hostname(hostname, &address.sin_addr);
-	address.sin_family = AF_INET;
-	address.sin_port = htons((unsigned short)dport);
-
-	sockfd = socket(address.sin_family, SOCK_STREAM, 0);
-	if (sockfd == SOCK_ERR) {
+	
+	/* Make a string of the port number that can be used with getaddrinfo */
+	sprintf(sport, "%d", dport);
+
+	/* Create hints for getaddrinfo saying that we want IPv4 and IPv6 TCP stream sockets */
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+	hints.ai_flags = AI_PASSIVE;
+
+	/* Call getaddrinfo and return SOCK_ERR if the call fails for some reason */
+	if ((status = getaddrinfo(hostname, sport, &hints, &remote)) != 0) {
 #ifndef DEBUGGER_FAIL_SILENTLY
 #if WIN32|WINNT
-		printf("create_debugger_socket(\"%s\", %d) socket: %d\n",
+		printf("create_debugger_socket(\"%s\", %d) getaddrinfo: %d\n",
 					hostname, dport, WSAGetLastError());
 #else
-		printf("create_debugger_socket(\"%s\", %d) socket: %s\n",
+		printf("create_debugger_socket(\"%s\", %d) getaddrinfo: %s\n",
 					hostname, dport, strerror(errno));
 #endif
 #endif
-		return -1;
+		return SOCK_ERR;
 	}
 
-	/* Put socket in non-blocking mode so we can use select for timeouts */
-	timeout.tv_sec = 0;
-	timeout.tv_usec = 200000;
+	/* Go through every returned IP address */
+	for (ptr = remote; ptr != NULL; ptr = ptr->ai_next) {
+		/* Try to create the socket. If the creation fails continue on with the next IP address in the list */
```

Comment:
You are correct. But at the same time maybe the IDEs that doesn't support IPv6 isn't listening on the port so the connection over IPv6 fails and xdebug then tries over IPv4. That is how I believe it would work, but can't of course not guarantee it because it's still an untested case. :-)